File: lucid/data/tokenizers/wordpiece.py
import unicodedata

from collections import Counter
from pathlib import Path
from typing import Any, Iterable

from lucid.data.tokenizers import Tokenizer, SpecialTokens

from lucid._backend._C.tokenizers.core import _C_WordPieceTokenizer


__all__ = ["WordPieceTokenizer", "WordPieceTokenizerFast"]


class WordPieceTokenizer(Tokenizer):
    def __init__(
        self,
        vocab: dict[str, int] | None = None,
        vocab_file: Path | str | None = None,
        unk_token: SpecialTokens | str = SpecialTokens.UNK,
        pad_token: SpecialTokens | str = SpecialTokens.PAD,
        bos_token: SpecialTokens | str | None = None,
        eos_token: SpecialTokens | str | None = None,
        lowercase: bool = True,
        wordpieces_prefix: str = "##",
        max_input_chars_per_word: int = 100,
        clean_text: bool = True,
        verbose: bool = False,
    ) -> None:
        if vocab is not None and vocab_file is not None:
            raise ValueError("Provide only one of 'vocab' or 'vocab_file'.")

        super().__init__(
            unk_token=unk_token,
            pad_token=pad_token,
            bos_token=bos_token,
            eos_token=eos_token,
        )

        if vocab_file is not None:
            vocab = self._load_vocab(vocab_file)
        if vocab is None:
            vocab = {}

        self.lowercase = lowercase
        self.wordpieces_prefix = wordpieces_prefix
        self.max_input_chars_per_word = max_input_chars_per_word
        self.clean_text = clean_text

        self.vocab: dict[str, int] = dict(vocab)
        self._ensure_special_tokens()
        self.ids_to_tokens = self._build_id_to_token_map(self.vocab, verbose=verbose)

    def fit(
        self,
        texts: Iterable[str],
        vocab_size: int,
        min_frequency: int = 2,
        verbose: bool = False,
    ) -> WordPieceTokenizer:
        trained_vocab = self._train_vocab(
            texts=texts,
            vocab_size=vocab_size,
            min_frequency=min_frequency,
        )
        self.vocab = trained_vocab
        unk_token = self.unk_token or SpecialTokens.UNK.value

        self.ids_to_tokens = self._build_id_to_token_map(
            self.vocab, unk_token=unk_token, verbose=verbose
        )
        return self

    @classmethod
    def train_from_iterator(
        cls,
        texts: Iterable[str],
        vocab_size: int,
        min_frequency: int = 2,
        verbose: bool = False,
        **kwargs,
    ) -> WordPieceTokenizer:
        tokenizer = cls(**kwargs)
        tokenizer.fit(
            texts, vocab_size=vocab_size, min_frequency=min_frequency, verbose=verbose
        )
        return tokenizer

    @property
    def vocab_size(self) -> int:
        return len(self.vocab)

    def tokenize(self, text: str) -> list[str]:
        if self.clean_text:
            text = self._clean_text(text)

        basic_tokens = self._basic_tokenize(text)
        tokens: list[str] = []
        for token in basic_tokens:
            tokens.extend(self._wordpiece_tokenize(token))
        return tokens

    def convert_tokens_to_ids(self, tokens: str | list[str]) -> int | list[int]:
        unk_id = self._unk_id()
        if isinstance(tokens, str):
            return self.vocab.get(tokens, unk_id)
        return [self.vocab.get(token, unk_id) for token in tokens]

    def convert_ids_to_tokens(self, ids: int | list[int]) -> str | list[str]:
        if isinstance(ids, int):
            return self._id_to_token(ids)
        return [self._id_to_token(i) for i in ids]

    def convert_tokens_to_string(self, tokens: list[str]) -> str:
        out = " ".join(tokens)
        out = out.replace(f" {self.wordpieces_prefix}", "")

        for punct in [".", ",", "!", "?", ";", ":", "%", ")", "]", "}"]:
            out = out.replace(f" {punct}", punct)
        for punct in ["(", "[", "{"]:
            out = out.replace(f"{punct} ", punct)

        return out.strip()

    def save_pretrained(self, save_directory: Path | str) -> list[str]:
        save_path = Path(save_directory)
        save_path.mkdir(parents=True, exist_ok=True)

        vocab_path = save_path / "vocab.txt"
        config_path = save_path / "tokenizer_config.json"

        with vocab_path.open("w", encoding="utf-8") as f:
            for token in self.ids_to_tokens:
                f.write(token + "\n")

        config = {
            "unk_token": self.unk_token,
            "pad_token": self.pad_token,
            "bos_token": self.bos_token,
            "eos_token": self.eos_token,
            "lowercase": self.lowercase,
            "wordpieces_prefix": self.wordpieces_prefix,
            "max_input_chars_per_word": self.max_input_chars_per_word,
            "clean_text": self.clean_text,
        }
        self._save_tokenizer_config(save_path, config)

        return [str(vocab_path), str(config_path)]

    @classmethod
    def from_pretrained(
        cls, pretrained_model_name_or_path: str | Path, **kwrags
    ) -> WordPieceTokenizer:
        path = Path(pretrained_model_name_or_path)
        if path.is_dir():
            vocab_file = path / "vocab.txt"
        else:
            vocab_file = path

        if not vocab_file.exists():
            raise FileNotFoundError(f"Cannot find vocabulary file: {vocab_file}")

        allowed = {
            "unk_token",
            "pad_token",
            "bos_token",
            "eos_token",
            "lowercase",
            "wordpieces_prefix",
            "max_input_chars_per_word",
            "clean_text",
        }
        init_kwargs = cls._load_tokenizer_config(path, allowed_keys=allowed)

        init_kwargs.update(kwrags)
        return cls(vocab_file=vocab_file, **init_kwargs)

    @staticmethod
    def _load_vocab(vocab_file: Path | str) -> dict[str, int]:
        vocab: dict[str, int] = {}
        with Path(vocab_file).open("r", encoding="utf-8") as f:
            for idx, line in enumerate(f):
                token = line.rstrip("\n")
                if token:
                    vocab[token] = idx
        return vocab

    def _ensure_special_tokens(self) -> None:
        for token in self.all_special_tokens:
            if token not in self.vocab:
                self.vocab[token] = len(self.vocab)

    def _train_vocab(
        self,
        texts: Iterable[str],
        vocab_size: int,
        min_frequency: int,
    ) -> dict[str, int]:
        special_tokens = self.all_special_tokens
        if vocab_size < len(special_tokens):
            raise ValueError(
                f"'vocab_size' must be >= number of special tokens ({len(special_tokens)})."
            )
        if min_frequency < 1:
            raise ValueError("'min_frequency' must be >= 1.")

        word_freq = self._build_word_frequency(texts)
        if not word_freq:
            raise ValueError("Cannot train WordPiece vocabulary from empty corpus.")

        word_splits, token_set = self._initialize_splits(word_freq)
        merged_tokens: list[str] = []

        target_size = vocab_size - len(special_tokens)
        while len(token_set) + len(merged_tokens) < target_size:
            best_pair = self._select_best_pair(word_splits, word_freq)
            if best_pair is None:
                break
            merged = self._merge_pair(best_pair, word_splits)
            if merged not in token_set and merged not in merged_tokens:
                merged_tokens.append(merged)

        vocab_tokens = list(dict.fromkeys(sorted(token_set) + merged_tokens))
        vocab_tokens = vocab_tokens[:target_size]

        full_vocab = special_tokens + vocab_tokens
        return {token: i for i, token in enumerate(full_vocab)}

    def _build_word_frequency(self, texts: Iterable[str]) -> Counter[str]:
        word_freq: Counter[str] = Counter()
        for text in texts:
            if self.clean_text:
                text = self._clean_text(text)
            for token in self._basic_tokenize(text):
                if token:
                    word_freq[token] += 1
        return word_freq

    def _initialize_splits(
        self, word_freq: Counter[str]
    ) -> tuple[dict[str, list[str]], set[str]]:
        word_splits: dict[str, list[str]] = {}
        token_set: set[str] = set()

        for word in word_freq:
            if not word:
                continue
            split = [word[0]] + [self.wordpieces_prefix + ch for ch in word[1:]]
            word_splits[word] = split
            token_set.update(split)

        return word_splits, token_set

    def _select_best_pair(
        self, word_splits: dict[str, list[str]], word_freq: Counter[str]
    ) -> tuple[str, str] | None:
        token_freq: Counter[str] = Counter()
        pair_freq: Counter[tuple[str, str]] = Counter()

        for word, split in word_splits.items():
            freq = word_freq[word]
            for token in split:
                token_freq[token] += freq
            for i in range(len(split) - 1):
                pair_freq[(split[i], split[i + 1])] += freq

        best_pair: tuple[str, str] | None = None
        best_score = -1.0
        best_pair_count = -1

        for pair, pfreq in pair_freq.items():
            if pfreq < 1:
                continue
            left, right = pair
            denom = token_freq[left] * token_freq[right]
            if denom == 0:
                continue
            score = pfreq / denom
            if (
                score > best_score
                or (score == best_score and pfreq > best_pair_count)
                or (
                    score == best_score
                    and pfreq == best_pair_count
                    and pair < (best_pair or pair)
                )
            ):
                best_score = score
                best_pair_count = pfreq
                best_pair = pair

        if best_pair is None or best_pair_count < 1:
            return None
        return best_pair

    def _merge_pair(
        self, pair: tuple[str, str], word_splits: dict[str, list[str]]
    ) -> str:
        left, right = pair
        if right.startswith(self.wordpieces_prefix):
            merged_token = left + right[len(self.wordpieces_prefix) :]
        else:
            merged_token = left + right

        for word, split in word_splits.items():
            if len(split) < 2:
                continue
            new_split: list[str] = []
            i = 0
            while i < len(split):
                if i < len(split) - 1 and split[i] == left and split[i + 1] == right:
                    new_split.append(merged_token)
                    i += 2
                else:
                    new_split.append(split[i])
                    i += 1
            word_splits[word] = new_split

        return merged_token

    def _unk_id(self) -> int:
        unk = self.unk_token
        if unk is None:
            raise ValueError("'unk_token' must not be None for WordPiece tokenizer.")
        if unk not in self.vocab:
            raise ValueError(f"Unknown token '{unk}' is not in vocabulary.")
        return self.vocab[unk]

    def _id_to_token(self, idx: int) -> str:
        if idx < 0 or idx >= len(self.ids_to_tokens):
            return self.unk_token or SpecialTokens.UNK.value
        token = self.ids_to_tokens[idx]
        return token if token else (self.unk_token or SpecialTokens.UNK.value)

    def _basic_tokenize(self, text: str) -> list[str]:
        tokens: list[str] = []
        for token in text.strip().split():
            if self.lowercase:
                token = token.lower()
            tokens.extend(self._split_on_punctuation(token))
        return tokens

    @staticmethod
    def _split_on_punctuation(token: str) -> list[str]:
        out: list[str] = []
        current: list[str] = []
        for ch in token:
            if WordPieceTokenizer._is_punctuation(ch):
                if current:
                    out.append("".join(current))
                    current = []
                out.append(ch)
            else:
                current.append(ch)
        if current:
            out.append("".join(current))
        return out

    def _wordpiece_tokenize(self, token: str) -> list[str]:
        if len(token) > self.max_input_chars_per_word:
            return [self.unk_token or SpecialTokens.UNK.value]

        chars = list(token)
        start = 0
        sub_tokens: list[str] = []

        while start < len(chars):
            end = len(chars)
            cur_substr: str | None = None
            while start < end:
                substr = "".join(chars[start:end])
                if start > 0:
                    substr = self.wordpieces_prefix + substr
                if substr in self.vocab:
                    cur_substr = substr
                    break
                end -= 1

            if cur_substr is None:
                return [self.unk_token or SpecialTokens.UNK.value]

            sub_tokens.append(cur_substr)
            start = end

        return sub_tokens

    @staticmethod
    def _is_whitespace(ch: str) -> bool:
        if ch in (" ", "\t", "\n", "\r"):
            return True
        return unicodedata.category(ch) == "Zs"

    @staticmethod
    def _is_control(ch: str) -> bool:
        if ch in ("\t", "\n", "\r"):
            return False
        return unicodedata.category(ch) in {"Cc", "Cf"}

    @staticmethod
    def _is_punctuation(ch: str) -> bool:
        cp = ord(ch)
        if (
            (33 <= cp <= 47)
            or (58 <= cp <= 64)
            or (91 <= cp <= 96)
            or (123 <= cp <= 126)
        ):
            return True
        return unicodedata.category(ch).startswith("P")

    @staticmethod
    def _clean_text(text: str) -> str:
        out: list[str] = []
        for ch in text:
            cp = ord(ch)
            if cp in (0, 0xFFFD) or WordPieceTokenizer._is_control(ch):
                continue
            if WordPieceTokenizer._is_whitespace(ch):
                out.append(" ")
            else:
                out.append(ch)
        return "".join(out)


class WordPieceTokenizerFast(Tokenizer):
    def __init__(
        self,
        vocab: dict[str, int] | None = None,
        vocab_file: Path | str | None = None,
        unk_token: SpecialTokens | str = SpecialTokens.UNK,
        pad_token: SpecialTokens | str = SpecialTokens.PAD,
        bos_token: SpecialTokens | str | None = None,
        eos_token: SpecialTokens | str | None = None,
        lowercase: bool = True,
        wordpieces_prefix: str = "##",
        max_input_chars_per_word: int = 100,
        clean_text: bool = True,
    ) -> None:
        super().__init__(
            unk_token=unk_token,
            pad_token=pad_token,
            bos_token=bos_token,
            eos_token=eos_token,
        )

        self.lowercase = lowercase
        self.wordpieces_prefix = wordpieces_prefix
        self.max_input_chars_per_word = max_input_chars_per_word
        self.clean_text = clean_text

        if _C_WordPieceTokenizer is None:
            raise ImportError(
                "Cannot import native C++ WordPiece backend "
                "'lucid._backend._C.tokenizers.core'. "
                "Build extensions first."
            )

        self._backend = _C_WordPieceTokenizer(
            vocab=vocab,
            vocab_file=Path(vocab_file) if vocab_file is not None else None,
            unk_token=self.unk_token,
            pad_token=self.pad_token,
            bos_token=self.bos_token,
            eos_token=self.eos_token,
            wordpieces_prefix=self.wordpieces_prefix,
            max_input_chars_per_word=self.max_input_chars_per_word,
            lowercase=self.lowercase,
            clean_text=self.clean_text,
        )

    @property
    def vocab_size(self) -> int:
        size = getattr(self._backend, "vocab_size")
        return int(size() if callable(size) else size)

    def tokenize(self, text: str) -> list[str]:
        return list(self._backend.tokenize(text))

    def convert_tokens_to_ids(self, tokens: str | list[str]) -> int | list[int]:
        if isinstance(tokens, str):
            return int(self._backend.convert_token_to_id(tokens))
        return [int(x) for x in self._backend.convert_tokens_to_ids(tokens)]

    def convert_ids_to_tokens(self, ids: int | list[int]) -> str | list[str]:
        if isinstance(ids, int):
            return str(self._backend.convert_id_to_token(ids))
        return [str(x) for x in self._backend.convert_ids_to_tokens(ids)]

    def convert_tokens_to_string(self, tokens: list[str]) -> str:
        return str(self._backend.convert_tokens_to_string(tokens))

    def fit(
        self,
        texts: Iterable[str],
        vocab_size: int,
        min_frequency: int = 2,
        **_: Any,
    ) -> WordPieceTokenizerFast:
        self._backend.fit(list(texts), int(vocab_size), int(min_frequency))
        return self

    @classmethod
    def train_from_iterator(
        cls,
        texts: Iterable[str],
        vocab_size: int,
        min_frequency: int = 2,
        **kwargs: Any,
    ) -> WordPieceTokenizerFast:
        tokenizer = cls(**kwargs)
        tokenizer.fit(texts, vocab_size=vocab_size, min_frequency=min_frequency)
        return tokenizer

    def save_pretrained(self, save_directory: Path | str) -> list[str]:
        save_path = Path(save_directory)
        save_path.mkdir(parents=True, exist_ok=True)

        vocab_path = save_path / "vocab.txt"
        with vocab_path.open("w", encoding="utf-8") as f:
            for idx in range(self.vocab_size):
                f.write(self.convert_ids_to_tokens(idx) + "\n")

        config = {
            "unk_token": self.unk_token,
            "pad_token": self.pad_token,
            "bos_token": self.bos_token,
            "eos_token": self.eos_token,
            "lowercase": self.lowercase,
            "wordpieces_prefix": self.wordpieces_prefix,
            "max_input_chars_per_word": self.max_input_chars_per_word,
            "clean_text": self.clean_text,
        }
        config_path = self._save_tokenizer_config(save_path, config)
        return [str(vocab_path), str(config_path)]

    @classmethod
    def from_pretrained(
        cls, pretrained_model_name_or_path: str | Path, **kwrags: Any
    ) -> WordPieceTokenizerFast:
        path = Path(pretrained_model_name_or_path)
        vocab_file = path / "vocab.txt" if path.is_dir() else path
        if not vocab_file.exists():
            raise FileNotFoundError(f"Cannot find vocabulary file: {vocab_file}")

        allowed = {
            "unk_token",
            "pad_token",
            "bos_token",
            "eos_token",
            "lowercase",
            "wordpieces_prefix",
            "max_input_chars_per_word",
            "clean_text",
        }
        init_kwargs = cls._load_tokenizer_config(path, allowed_keys=allowed)
        init_kwargs.update(kwrags)
        return cls(vocab_file=vocab_file, **init_kwargs)

File: lucid/_C/tokenizers/wordpiece.cpp
#include "tokenizers.hpp"

#include <algorithm>
#include <cctype>
#include <fstream>
#include <stdexcept>
#include <unordered_map>
#include <utility>

namespace lucid::tokenizers::fast {
    WordPieceTokenizer::WordPieceTokenizer(
        std::optional<Vocab> vocab,
        std::optional<std::filesystem::path> vocab_file,

        std::string unk_token,
        std::string pad_token,

        std::optional<std::string> bos_token,
        std::optional<std::string> eos_token,

        std::string wordpieces_prefix,
        std::size_t max_input_chars_per_word,

        bool lowercase,
        bool clean_text
    ) {
        if (vocab.has_value() && vocab_file.has_value()) {
            throw std::invalid_argument("Provide only one of 'vocab' or 'vocab_file'.");
        }

        set_special_tokens(
            std::move(unk_token),
            std::move(pad_token),
            std::move(bos_token),
            std::move(eos_token)
        );

        if (vocab_file.has_value()) vocab_ = load_vocab(*vocab_file);
        else if (vocab.has_value()) vocab_ = std::move(*vocab);
        else vocab_.clear();

        wordpieces_prefix_ = std::move(wordpieces_prefix);
        max_input_chars_per_word_ = max_input_chars_per_word;
        lowercase_ = lowercase;
        clean_text_ = clean_text;

        ensure_special_tokens();
        rebuild_id_to_tokens();
    }

    std::size_t WordPieceTokenizer::vocab_size() const {
        return vocab_.size();
    }

    std::vector<std::string> WordPieceTokenizer::tokenize(std::string_view text) const {
        std::string input(text);
        if (clean_text_) input = detail::clean_text(input);

        const auto basic_tokens = detail::basic_tokenize(input, lowercase_);
        std::vector<std::string> out;

        for(const auto& tok : basic_tokens) {
            const auto wp = wordpiece_tokenize(tok);
            out.insert(out.end(), wp.begin(), wp.end());
        }
        return out;
    }

    int32_t WordPieceTokenizer::convert_token_to_id(std::string_view token) const {
        auto it = vocab_.find(std::string(token));
        if (it != vocab_.end()) return it->second;
        return unk_id();
    }

    std::vector<int32_t> WordPieceTokenizer::convert_tokens_to_ids(
        const std::vector<std::string>& tokens
    ) const {
        std::vector<int32_t> out;
        out.reserve(tokens.size());

        for (const auto& t : tokens) {
            out.push_back(convert_token_to_id(t));
        }
        return out;
    }

    std::string WordPieceTokenizer::convert_id_to_token(int32_t id) const {
        return id_to_token_impl(id);
    }

    std::vector<std::string> WordPieceTokenizer::convert_ids_to_tokens(
        const std::vector<int32_t>& ids
    ) const {
        std::vector<std::string> out;
        out.reserve(ids.size());

        for (int32_t id : ids){
            out.push_back(convert_id_to_token(id));
        }
        return out;
    }

    std::string WordPieceTokenizer::convert_tokens_to_string(
        const std::vector<std::string>& tokens
    ) const {
        std::string out;
        for (std::size_t i = 0; i < tokens.size(); ++i) {
            if (i > 0) out += " ";
            out += tokens[i];
        }

        const std::string pref = " " + wordpieces_prefix_;
        std::size_t pos = 0;
        while ((pos = out.find(pref, pos)) != std::string::npos) {
            out.erase(pos, pref.size());
        }

        const std::vector<std::string> right_attach = {
            ".", ",", "!", "?", ";", ":", "%", ")", "]", "}"
        };
        for (const auto& p : right_attach) {
            const std::string needle = " " + p;
            std::size_t rpos = 0;
            while ((rpos = out.find(needle, rpos)) != std::string::npos) {
                out.replace(rpos, needle.size(), p);
                rpos += p.size();
            }
        }

        const std::vector<std::string> left_attach = {"(", "[", "{"};
        for (const auto& p : left_attach) {
            const std::string needle = p + " ";
            std::size_t lpos = 0;
            while ((lpos = out.find(needle, lpos)) != std::string::npos) {
                out.replace(lpos, needle.size(), p);
                lpos += p.size();
            }
        }

        while (!out.empty() && std::isspace(static_cast<unsigned char>(out.front()))) {
            out.erase(out.begin());
        }
        while (!out.empty() && std::isspace(static_cast<unsigned char>(out.back()))) {
            out.pop_back();
        }

        std::string deduped;
        deduped.reserve(out.size());
        for (char ch : out) {
            if (!deduped.empty()) {
                const char prev = deduped.back();
                const bool same = (ch == prev);
                const bool punct = (ch == '.' || ch == '!' || ch == '?');
                if (same && punct) continue;
            }
            deduped.push_back(ch);
        }

        return deduped;
    }

    WordPieceTokenizer& WordPieceTokenizer::fit(
        const std::vector<std::string>& texts,
        std::size_t target_vocab_size,
        std::size_t min_frequency
    ) {
        vocab_ = train_vocab(texts, target_vocab_size, min_frequency);
        rebuild_id_to_tokens();
        return *this;
    }

    WordPieceTokenizer::Vocab WordPieceTokenizer::load_vocab(
        const std::filesystem::path& vocab_file
    ) {
        std::ifstream fin(vocab_file);
        if (!fin.is_open()) {
            throw std::runtime_error(
                "Failed to open vocabulary file: " + vocab_file.string()
            );
        }
        Vocab vocab;
        std::string line;
        int32_t idx = 0;
        while (std::getline(fin, line)) {
            if (!line.empty()) vocab[line] = idx;
            ++idx;
        }
        return vocab;
    }

    void WordPieceTokenizer::ensure_special_tokens() {
        const auto specials = all_special_tokens();
        for (const auto& token : specials) {
            if (vocab_.find(token) == vocab_.end()) {
                vocab_[token] = static_cast<int32_t>(vocab_.size());
            }
        }
    }

    void WordPieceTokenizer::rebuild_id_to_tokens() {
        int32_t max_id = -1;
        for (const auto& [token, id] : vocab_) {
            (void)token;
            if (id < 0) {
                throw std::runtime_error(
                    "Vocabulary indices must be non-negative integers."
                );
            }
            if (id > max_id) max_id = id;
        }
        if (max_id < 0) {
            ids_to_tokens_.clear();
            return;
        }

        ids_to_tokens_.assign(static_cast<std::size_t>(max_id) + 1, "");
        for (const auto& [token, id] : vocab_) {
            auto& slot = ids_to_tokens_[static_cast<std::size_t>(id)];
            if (!slot.empty()) {
                throw std::runtime_error(
                    "Duplicate token index detected: " + std::to_string(id)
                );
            }
            slot = token;
        }

        for (auto& token : ids_to_tokens_) {
            if (token.empty()) {
                token = unk_token_;
            }
        }
    }

    int32_t WordPieceTokenizer::unk_id() const {
        const auto it = vocab_.find(unk_token_);
        if (it == vocab_.end()) {
            throw std::runtime_error(
                "Unknown token '" + unk_token_ + "' is not in vocabulary."
            );
        }
        return it->second;
    }

    std::string WordPieceTokenizer::id_to_token_impl(int32_t id) const {
        if (id < 0 || static_cast<std::size_t>(id) >= ids_to_tokens_.size() ){
            return unk_token_;
        }
        const auto& tok = ids_to_tokens_[static_cast<std::size_t>(id)];
        return tok.empty() ? unk_token_ : tok;
    }

    std::vector<std::string> WordPieceTokenizer::wordpiece_tokenize(
        const std::string& token
    ) const {
        if (token.size() > max_input_chars_per_word_) return {unk_token_};

        std::vector<std::string> out;
        std::size_t start = 0;
        while (start < token.size()) {
            std::size_t end = token.size();
            bool found = false;
            std::string best;

            while (start < end) {
                std::string sub = token.substr(start, end - start);
                if (start > 0) sub = wordpieces_prefix_ + sub;
                
                if (vocab_.find(sub) != vocab_.end()) {
                    best = std::move(sub);
                    found = true;
                    break;
                }
                --end;
            }
            if (!found) return {unk_token_};

            out.push_back(std::move(best));
            start = end;
        }
        return out;
    }

    WordPieceTokenizer::Vocab WordPieceTokenizer::train_vocab(
        const std::vector<std::string>& texts,
        std::size_t target_vocab_size,
        std::size_t min_frequency
    ) const {
        const auto special_tokens = all_special_tokens();
        if (target_vocab_size < special_tokens.size()) {
            throw std::invalid_argument(
                "'vocab_size' must be >= number of special tokens ("
                + std::to_string(special_tokens.size()) + ")."
            );
        }
        if (min_frequency < 1) {
            throw std::invalid_argument(
                "'min_frequency' must be >= 1."
            );
        }
        const auto word_freq = build_word_frequency(texts);
        if (word_freq.empty()) throw std::invalid_argument(
            "Cannot train WordPiece vocabulary from empty corpus."
        );

        auto [word_splits, token_set] = initialize_splits(word_freq);
        std::vector<std::string> merged_tokens;
        std::unordered_map<std::string, bool> merged_seen;

        const std::size_t token_target = target_vocab_size - special_tokens.size();

        while ((token_set.size() + merged_tokens.size()) < token_target) {
            const auto best_pair = select_best_pair(word_splits, word_freq);
            if (!best_pair.has_value()) break;

            const std::string merged = merge_pair(*best_pair, word_splits);
            if (
                token_set.find(merged) == token_set.end() 
                && merged_seen.find(merged) == merged_seen.end()
            ) {
                merged_tokens.push_back(merged);
                merged_seen[merged] = true;
            }
        }

        std::vector<std::string> token_list;
        token_list.reserve(token_set.size());

        for (const auto& [tok, seen] : token_set) {
            if (seen) token_list.push_back(tok);
        }
        std::sort(token_list.begin(), token_list.end());
        token_list.insert(token_list.end(), merged_tokens.begin(), merged_tokens.end());

        std::vector<std::string> vocab_tokens;
        vocab_tokens.reserve(std::min(token_target, token_list.size()));

        std::unordered_map<std::string, bool> seen;
        for (const auto& tok : token_list) {
            if (seen.find(tok) != seen.end()) continue;
            seen[tok] = true;

            vocab_tokens.push_back(tok);
            if (vocab_tokens.size() >= token_target) break;
        }

        std::vector<std::string> full_vocab;
        
        full_vocab.reserve(special_tokens.size() + vocab_tokens.size());
        full_vocab.insert(full_vocab.end(), special_tokens.begin(), special_tokens.end());
        full_vocab.insert(full_vocab.end(), vocab_tokens.begin(), vocab_tokens.end());

        Vocab trained;
        for (std::size_t i = 0; i < full_vocab.size(); ++i) {
            trained[full_vocab[i]] = static_cast<int32_t>(i);
        }
        return trained;
    }

    std::unordered_map<std::string, std::size_t> WordPieceTokenizer::build_word_frequency(
        const std::vector<std::string>& texts
    ) const {
        std::unordered_map<std::string, std::size_t> word_freq;
        for (const auto& raw_text: texts) {
            std::string text = raw_text;
            if (clean_text_) text = detail::clean_text(text);

            const auto tokens = detail::basic_tokenize(text, lowercase_);
            for (const auto& token : tokens) {
                if (!token.empty()) ++word_freq[token];
            }
        }
        return word_freq;
    }

    std::pair<
        std::unordered_map<std::string, std::vector<std::string>>,
        std::unordered_map<std::string, bool>
    > WordPieceTokenizer::initialize_splits(
        const std::unordered_map<std::string, std::size_t>& word_freq
    ) const {
        std::unordered_map<std::string, std::vector<std::string>> word_splits;
        std::unordered_map<std::string, bool> token_set;

        for (const auto& [word, freq] : word_freq) {
            (void)freq;
            if (word.empty()) continue;

            std::vector<std::string> split;
            split.reserve(word.size());
            split.emplace_back(1, word[0]);

            for (std::size_t i = 1; i < word.size(); ++i) {
                split.push_back(wordpieces_prefix_ + std::string(1, word[i]));
            }

            word_splits[word] = split;
            for (const auto& tok : split) token_set[tok] = true;
        }
        return {std::move(word_splits), std::move(token_set)};
    }

    std::optional<std::pair<std::string, std::string>> WordPieceTokenizer::select_best_pair(
        const std::unordered_map<std::string, std::vector<std::string>>& word_splits,
        const std::unordered_map<std::string, std::size_t>& word_freq
    ) const {
        std::unordered_map<std::string, std::size_t> token_freq;
        std::unordered_map<detail::TokenPair, std::size_t, detail::TokenPairHash> pair_freq;

        for (const auto& [word, split] : word_splits) {
            auto wf_it = word_freq.find(word);
            if (wf_it == word_freq.end()) continue;

            const std::size_t freq = wf_it->second;
            for (const auto& tok : split) {
                token_freq[tok] += freq;
            }
            for (std::size_t i = 0; i + 1 < split.size(); ++i) {
                pair_freq[{split[i], split[i + 1]}] += freq;
            }
        }

        bool has_best = false;
        std::pair<std::string, std::string> best_pair;
        double best_score = -1.0;
        std::size_t best_pair_count = 0;

        for (const auto& [pair, pfreq]: pair_freq) {
            if (pfreq < 1) continue;

            const auto& left = pair.first;
            const auto& right = pair.second;

            auto l_it = token_freq.find(left);
            auto r_it = token_freq.find(right);
            if (l_it == token_freq.end() || r_it == token_freq.end()) continue;

            const std::size_t denom = l_it->second * r_it->second;
            if (denom == 0) continue;

            const double score = static_cast<double>(pfreq) / static_cast<double>(denom);

            if (
                !has_best
                || score > best_score
                || (score == best_score && pfreq > best_pair_count)
                || (score == best_score && pfreq == best_pair_count && pair < best_pair)
            ) {
                has_best = true;
                best_pair = pair;
                best_score = score;
                best_pair_count = pfreq;
            }
        }
        if (!has_best || best_pair_count < 1) return std::nullopt;
        return best_pair;
    }

    std::string WordPieceTokenizer::merge_pair(
        const std::pair<std::string, std::string>& pair,
        std::unordered_map<std::string, std::vector<std::string>>& word_splits
    ) const {
        const auto& left = pair.first;
        const auto& right = pair.second;

        std::string merged_token;
        if (right.rfind(wordpieces_prefix_, 0) == 0) {
            merged_token = left + right.substr(wordpieces_prefix_.size());
        } else {
            merged_token = left + right;
        }

        for (auto& [word, split] : word_splits) {
            (void)word;
            if (split.size() < 2) continue;

            std::vector<std::string> new_split;
            new_split.reserve(split.size());

            std::size_t i = 0;
            while (i < split.size()) {
                if (i + 1 < split.size() && split[i] == left && split[i + 1] == right) {
                    new_split.push_back(merged_token);
                    i += 2;
                } else {
                    new_split.push_back(split[i]);
                    ++i;
                }
            }
            split = std::move(new_split);
        }
        return merged_token;
    }
}
